### PR TITLE
Collision detection

### LINKED ISSUE #8: Implement collision detection and resolution
- Add Spatial Hash.
 - Add global Collision Handler component.
 - Add collider components (circle, box, etc).
 - Add helper methods for checking collisions (ray casts, etc).

### PR DESCRIPTION
**Pull request by lunacys**
Closes #8

Implemented: 

 - [x] Collider shapes (`ColliderShape`):
   - [x] Circle
   - [x] Rectangle (Box)
 - [x] Collision detection between different shapes (`collider.collides_with()` or `collider.collides_with_motion()`)
 - [x] Overlapping detection (`collider.overlaps()`)
 - [x] Helper structs:
   - [x] `CollisionResult` - contains info about collider, collision normal vector and minimum translation vector
   - [x] `Ray2D` - represents a 2D ray
   - [x] `RaycastHit` - represents a raycast hit result
   - [x] `Rect` - `f32` rectangle with {x,y} being the top-left corner
 - [x] Spatial Hash for optimization purposes
   - [x] Adding/removing colliders
   - [x] Correct handling of collider registration - currently it registers before the position is fully set, which causes bugs
 - [x]  Correct collision resolution
   - [x] Enemies should avoid bumping into each other - buggy
   - [x] Use the new collision detection for checking for player-enemy collisions (draining health)

### CODE REVIEW

File: src/collisions.rs [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+                res.y = self.bottom();
+                edge_normal.y = 1.0;
+            } else if min == dl {
+                res.x = self.left();
+                edge_normal.x = -1.0;
+            } else {
+                res.x = self.right();
+                edge_normal.x = 1.0;
+            }
+        } else {
```
Comment (Djutifai) [inline src/collisions.rs:396]: maybe we should remove unneeded else so there will and just return our values in the upper if and do downward logic without else brackets?

File: src/collisions.rs [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+    }
+
+    fn ray_intersects(&self, ray: &Ray2D) -> Option<f32> {
+        let mut distance = 0.0;
+        let mut max = f32::MAX;
+
+        if ray.direction.x.abs() < 1e-06 {
+            if ray.start.x < self.x || ray.start.x > self.x + self.width {
+                return None;
+            }
```
Comment (Djutifai) [inline src/collisions.rs:421]:
```
if ray.direction.x.abs() < 1e-06 &&
             ray.start.x < self.x || ray.start.x > self.x + self.width {
             return None;
             }
```

?

File: src/collisions.rs [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+        } else {
+            let inv_x = 1.0 / ray.direction.x;
+            let mut left = (self.x - ray.start.x) * inv_x;
+            let mut right = (self.x + self.width - ray.start.x) * inv_x;
+            if left > right {
+                let tmp = left;
+                left = right;
+                right = tmp;
+            }
```
Comment (Djutifai) [inline src/collisions.rs:430]:
```
if left > right {
    mem::swap(&mut left, &mut right);
}
```

File: src/collisions.rs [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+                top = bottom;
+                bottom = tmp;
+            }
+
+            distance = top.max(distance);
+            max = bottom.min(max);
+            if distance > max {
+                return None;
+            }
+        }
```
Comment (Djutifai) [inline src/collisions.rs:458]: we probably should move it in function probably because it is the same 20 lines of code as for x direction

**Review by Djutifai — CHANGES_REQUESTED** [review]

File: src/collisions.rs [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+            } else if min == db {
+                res.y = self.bottom();
+                edge_normal.y = 1.0;
+            } else if min == dl {
+                res.x = self.left();
+                edge_normal.x = -1.0;
+            } else {
+                res.x = self.right();
+                edge_normal.x = 1.0;
+            }
```
[the remark this answers is not in the document]
Comment (lunacys) [reply src/collisions.rs:400]: We cannot use `match` here

File: src/collisions/spatial_hash.rs [hunk 6]
```diff
@@ -0,0 +1,79 @@
+/* use bevy::{
```
[the remark this answers is not in the document]
Comment (lunacys) [reply src/collisions/spatial_hash.rs]: Removed

File: src/enemy.rs [hunk 7]
```diff
@@ -319,13 +317,20 @@ fn spawn(
...
-                                    host: SteeringHost {
-                                        position: Vec2::new(m_x, m_y),
+                                    position: Position(Vec2::new(m_x, m_y)),
+                                    physics_params: PhysicsParams {
                                         max_velocity: spawner.max_velocity,
                                         max_force: spawner.max_force,
                                         mass: spawner.mass,
                                         ..default()
                                     },
+                                    ..default() /*host: SteeringHost {
```
Comment (Sereja313) [inline src/enemy.rs:327]: useless comment

File: src/steering.rs [hunk 8]
```diff
@@ -196,10 +278,12 @@ fn update_translation(mut host: Query<(&mut Transform, &SteeringHost)>) {
     }
 }
 
-fn update_position(time: Res<Time>, mut host: Query<&mut SteeringHost>) {
-    for mut host in &mut host {
-        let dt = host.cur_velocity * time.delta_seconds();
-        host.position += dt;
+fn update_position(time: Res<Time>, mut host: Query<(&mut SteeringHost, Entity)>) {
+    for (mut host, _entity) in &mut host {
```
[the remark this answers is not in the document]
Comment (lunacys) [reply src/steering.rs]: Fixed

File: src/collisions.rs [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+            if res.x == self.right() {
+                edge_normal.x = 1.0;
+            }
+            if res.y == self.top() {
+                edge_normal.y = -1.0;
+            }
+            if res.y == self.bottom() {
+                edge_normal.y = 1.0;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (lunacys) [reply src/collisions.rs:415]: Here we're setting the coordinates independently from each other

File: src/collisions.rs (hunk 2 above)
Comment (lunacys) [reply src/collisions.rs:427]: Your alternative is much harder to read imo

File: src/enemy.rs (hunk 7 above)
Comment (lunacys) [reply src/enemy.rs]: Fixed

File: src/collisions.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Djutifai) [reply src/collisions.rs:415]: whoops, yeah

File: src/collisions/shapes.rs [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+                        let depth = sum_of_radii - dist_sqr.sqrt();
+                        res.min_translation = -depth * res.normal;
+                        res.point = second_pos + res.normal * r2;
+
+                        return Some(res);
+                    }
+
+                    None
+                }
+                ColliderShapeType::Box { .. } => panic!("second: expected circle, bot box"),
```
Comment (Sereja313) [inline src/collisions/shapes.rs:78]: typo

File: src/collisions.rs (hunk 1 above)
Comment (Djutifai) [reply src/collisions.rs:402]: after if make return, remove else. else is unneeded in this case imo

Comment (Djutifai) [reply src/collisions.rs:402]:
```
if ... {
   ...
   ...
   return ...
}
  ...
  ...
  return ...
  ```

File: src/enemy.rs [hunk 11]
```diff
@@ -30,31 +37,105 @@ Die. [V]
...
+    enemies: Query<(With<Enemy>, Without<Player>)>,
+    mut text: Query<&mut Text, Without<FpsText>>,
+) {
+    if let Ok(mut text) = text.get_single_mut() {
+        text.sections[0].value = format!("Capybaras: {}", enemies.iter().count());
+    }
+}
+
+#[derive(Component)]
+struct FpsText;
```
Comment (Sereja313) [inline src/enemy.rs:117]: why inside `enemy.rs`?

File: src/collisions.rs (hunk 1 above)
Comment (lunacys) [reply src/collisions.rs:402]: We need an `else` here

File: src/enemy.rs (hunk 11 above)
Comment (lunacys) [reply src/enemy.rs:117]: for debugging purposes, probably should move it anywhere else

File: src/collisions/shapes.rs (hunk 10 above)
Comment (lunacys) [reply src/collisions/shapes.rs]: Fixed

File: src/enemy.rs (hunk 11 above)
Comment (lunacys) [reply src/enemy.rs]: Fixed, now it is in the `debug` module

File: src/collisions.rs (hunk 4 above)
Comment (lunacys) [reply src/collisions.rs]: Fixed

File: src/collisions.rs (hunk 3 above)
Comment (lunacys) [reply src/collisions.rs]: Fixed

File: src/collisions.rs [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+
+        if max.y.abs() < min_dist {
+            min_dist = max.y.abs();
+            bounds_point.x = 0.0;
+            bounds_point.y = max.y;
+        }
+
+        if self.y.abs() < min_dist {
+            // min_dist = self.y.abs();
```
Comment (Djutifai) [inline src/collisions.rs:365]: ?

File: src/collisions.rs (hunk 1 above)
Comment (Djutifai) [reply src/collisions.rs:402]: why

Comment (lunacys) [reply src/collisions.rs]: Fixed

File: src/collisions.rs (hunk 12 above)
Comment (lunacys) [reply src/collisions.rs]: Fixed

**Review by Djutifai — APPROVED** [review]
lgtm

**Pull request merged**